src/gather.jl

```diff
@@ -22,20 +25,21 @@ or multiple `dst` columns.
 
 See [`gather`](@ref) for an allocating version.
 """
+gather!(dst::AbstractArray, src::AbstractArray, idx::AbstractArray{<:IntOrIntTuple}) = gather!(dst, src, CartesianIndex.(idx))
+
 function gather!(dst::AbstractArray{Tdst,Ndst}, 
                  src::AbstractArray{Tsrc,Nsrc}, 
-                 idx::AbstractArray{Tidx, Nidx}) where 
-                    {Tdst, Tsrc, Ndst, Nsrc, Nidx, Tidx <: IntOrIntTuple}
+                 idx::AbstractArray{CartesianIndex{M},Nidx}) where 
+                    {Tdst, Tsrc, Ndst, Nsrc, M, Nidx}
 
-    M = typelength(Tidx)
-    d = Ndst - Nidx 
+    d = Ndst - Nidx
     d == Nsrc - M || throw(ArgumentError("Incompatible input shapes."))
     size(dst)[1:d] == size(src)[1:d] || throw(ArgumentError("Incompatible input shapes."))
     size(dst)[d+1:end] == size(idx) || throw(ArgumentError("Incompatible input shapes."))
 
     colons = ntuple(i -> Colon(), d)
     for k in CartesianIndices(idx)
-        view(dst, colons..., k) .= view(src, colons..., idx[k]...)
+        view(dst, colons..., k) .= view(src, colons..., idx[k])
```

Comment:
Maybe we can define
```julia
_view(src, colons, idx::AbstractArray{CartesianIndex}, k) = view(src, colons..., idx[k])
_view(src, colons, idx, k) = view(src, colons..., idx[k]...)
```
and use it here instead of converting everything to CartesianIndexes (see comment above)

Reply:
Best to avoid unnecessary indirection

Reply:
it's not unnecessary, I'm suggesting it in order to avoid a copy of the indexes, see comment above

Reply:
The compiler can avoid the complete copy if it can determine the types

Reply:
I don't think the compiler can avoid producing a copy when doing `CartesianIndex.(idx)`

Reply:
I checked the implementation of `CartesianIndex`. It contains a NTuple as a field, so it is surely that converting to a `CartesianIndex` will have new allocation or encapsulation.

Reply:
Do we need to loop over the CartesianIndices? Can't we use them directly?

Reply:
`CartesianIndices` instead of `AbstractArray{CartesianIndex}` ?

Reply:
`CartesianIndices` requires the dimensions of both inner `CartesianIndex` and outer `AbstractArray` to be the same. You may check it here: `struct CartesianIndices{N,R<:NTuple{N,OrdinalRange{Int, Int}}} <: AbstractArray{CartesianIndex{N},N}`. However, it is not required to be the same in scatter and gather operations. You may find a test case of `Array{NTuple{1},2}` in my test case.